File: Vanilla/sources/zomboid/zombie/iso/areas/isoregion/regions/IsoChunkRegion.java
package zombie.iso.areas.isoregion.regions;

import java.util.ArrayList;
import zombie.core.Color;
import zombie.core.Core;
import zombie.iso.areas.isoregion.IsoRegions;

public final class IsoChunkRegion implements IChunkRegion {
   private final IsoRegionManager manager;
   private boolean isInPool = false;
   private Color color;
   private int ID;
   private byte zLayer;
   private byte squareSize = 0;
   private byte roofCnt = 0;
   private byte chunkBorderSquaresCnt = 0;
   private final boolean[] enclosed = new boolean[4];
   private boolean enclosedCache = true;
   private final ArrayList connectedNeighbors = new ArrayList();
   private final ArrayList allNeighbors = new ArrayList();
   private boolean isDirtyEnclosed = false;
   private IsoWorldRegion isoWorldRegion;

   public int getID() {
      return this.ID;
   }

   public int getSquareSize() {
      return this.squareSize;
   }

   public Color getColor() {
      return this.color;
   }

   public int getzLayer() {
      return this.zLayer;
   }

   public IsoWorldRegion getIsoWorldRegion() {
      return this.isoWorldRegion;
   }

   public void setIsoWorldRegion(IsoWorldRegion var1) {
      this.isoWorldRegion = var1;
   }

   protected boolean isInPool() {
      return this.isInPool;
   }

   protected IsoChunkRegion(IsoRegionManager var1) {
      this.manager = var1;
   }

   protected void init(int var1, int var2) {
      this.isInPool = false;
      this.ID = var1;
      this.zLayer = (byte)var2;
      this.resetChunkBorderSquaresCnt();
      if (this.color == null) {
         this.color = this.manager.getColor();
      }

      this.squareSize = 0;
      this.roofCnt = 0;
      this.resetEnclosed();
   }

   protected IsoChunkRegion reset() {
      this.isInPool = true;
      this.unlinkNeighbors();
      IsoWorldRegion var1 = this.unlinkFromIsoWorldRegion();
      if (var1 != null && var1.size() <= 0) {
         if (Core.bDebug) {
            throw new RuntimeException("ChunkRegion.reset IsoChunkRegion has IsoWorldRegion with 0 members.");
         }

         this.manager.releaseIsoWorldRegion(var1);
         IsoRegions.warn("ChunkRegion.reset IsoChunkRegion has IsoWorldRegion with 0 members.");
      }

      this.resetChunkBorderSquaresCnt();
      this.ID = -1;
      this.squareSize = 0;
      this.roofCnt = 0;
      this.resetEnclosed();
      return this;
   }

   public IsoWorldRegion unlinkFromIsoWorldRegion() {
      if (this.isoWorldRegion != null) {
         IsoWorldRegion var1 = this.isoWorldRegion;
         this.isoWorldRegion.removeIsoChunkRegion(this);
         this.isoWorldRegion = null;
         return var1;
      } else {
         return null;
      }
   }

   public int getRoofCnt() {
      return this.roofCnt;
   }

   public void addRoof() {
      ++this.roofCnt;
      if (this.roofCnt > this.squareSize) {
         IsoRegions.warn("ChunkRegion.addRoof roofCount exceed squareSize.");
         this.roofCnt = this.squareSize;
      } else {
         if (this.isoWorldRegion != null) {
            this.isoWorldRegion.addRoof();
         }

      }
   }

   public void resetRoofCnt() {
      if (this.isoWorldRegion != null) {
         this.isoWorldRegion.removeRoofs(this.roofCnt);
      }

      this.roofCnt = 0;
   }

   public void addSquareCount() {
      ++this.squareSize;
   }

   public int getChunkBorderSquaresCnt() {
      return this.chunkBorderSquaresCnt;
   }

   public void addChunkBorderSquaresCnt() {
      ++this.chunkBorderSquaresCnt;
   }

   protected void removeChunkBorderSquaresCnt() {
      --this.chunkBorderSquaresCnt;
      if (this.chunkBorderSquaresCnt < 0) {
         this.chunkBorderSquaresCnt = 0;
      }

   }

   protected void resetChunkBorderSquaresCnt() {
      this.chunkBorderSquaresCnt = 0;
   }

   private void resetEnclosed() {
      for(byte var1 = 0; var1 < 4; ++var1) {
         this.enclosed[var1] = true;
      }

      this.isDirtyEnclosed = false;
      this.enclosedCache = true;
   }

   public void setEnclosed(byte var1, boolean var2) {
      this.isDirtyEnclosed = true;
      this.enclosed[var1] = var2;
   }

   protected void setDirtyEnclosed() {
      this.isDirtyEnclosed = true;
      if (this.isoWorldRegion != null) {
         this.isoWorldRegion.setDirtyEnclosed();
      }

   }

   public boolean getIsEnclosed() {
      if (!this.isDirtyEnclosed) {
         return this.enclosedCache;
      } else {
         this.isDirtyEnclosed = false;
         this.enclosedCache = true;

         for(byte var1 = 0; var1 < 4; ++var1) {
            if (!this.enclosed[var1]) {
               this.enclosedCache = false;
            }
         }

         if (this.isoWorldRegion != null) {
            this.isoWorldRegion.setDirtyEnclosed();
         }

         return this.enclosedCache;
      }
   }

   public ArrayList getConnectedNeighbors() {
      return this.connectedNeighbors;
   }

   public void addConnectedNeighbor(IsoChunkRegion var1) {
      if (var1 != null) {
         if (!this.connectedNeighbors.contains(var1)) {
            this.connectedNeighbors.add(var1);
         }

      }
   }

   protected void removeConnectedNeighbor(IsoChunkRegion var1) {
      this.connectedNeighbors.remove(var1);
   }

   public int getNeighborCount() {
      return this.allNeighbors.size();
   }

   protected ArrayList getAllNeighbors() {
      return this.allNeighbors;
   }

   public void addNeighbor(IsoChunkRegion var1) {
      if (var1 != null) {
         if (!this.allNeighbors.contains(var1)) {
            this.allNeighbors.add(var1);
         }

      }
   }

   protected void removeNeighbor(IsoChunkRegion var1) {
      this.allNeighbors.remove(var1);
   }

   protected void unlinkNeighbors() {
      IsoChunkRegion var1;
      int var2;
      for(var2 = 0; var2 < this.connectedNeighbors.size(); ++var2) {
         var1 = (IsoChunkRegion)this.connectedNeighbors.get(var2);
         var1.removeConnectedNeighbor(this);
      }

      this.connectedNeighbors.clear();

      for(var2 = 0; var2 < this.allNeighbors.size(); ++var2) {
         var1 = (IsoChunkRegion)this.allNeighbors.get(var2);
         var1.removeNeighbor(this);
      }

      this.allNeighbors.clear();
   }

   public ArrayList getDebugConnectedNeighborCopy() {
      ArrayList var1 = new ArrayList();
      if (this.connectedNeighbors.size() == 0) {
         return var1;
      } else {
         var1.addAll(this.connectedNeighbors);
         return var1;
      }
   }

   public boolean containsConnectedNeighbor(IsoChunkRegion var1) {
      return this.connectedNeighbors.contains(var1);
   }

   public boolean containsConnectedNeighborID(int var1) {
      if (this.connectedNeighbors.size() == 0) {
         return false;
      } else {
         for(int var3 = 0; var3 < this.connectedNeighbors.size(); ++var3) {
            IsoChunkRegion var2 = (IsoChunkRegion)this.connectedNeighbors.get(var3);
            if (var2.getID() == var1) {
               return true;
            }
         }

         return false;
      }
   }

   public IsoChunkRegion getConnectedNeighborWithLargestIsoWorldRegion() {
      if (this.connectedNeighbors.size() == 0) {
         return null;
      } else {
         IsoWorldRegion var1 = null;
         IsoChunkRegion var2 = null;

         for(int var4 = 0; var4 < this.connectedNeighbors.size(); ++var4) {
            IsoChunkRegion var3 = (IsoChunkRegion)this.connectedNeighbors.get(var4);
            IsoWorldRegion var5 = var3.getIsoWorldRegion();
            if (var5 != null && (var1 == null || var5.getSquareSize() > var1.getSquareSize())) {
               var1 = var5;
               var2 = var3;
            }
         }

         return var2;
      }
   }

   protected IsoChunkRegion getFirstNeighborWithIsoWorldRegion() {
      if (this.connectedNeighbors.size() == 0) {
         return null;
      } else {
         for(int var2 = 0; var2 < this.connectedNeighbors.size(); ++var2) {
            IsoChunkRegion var1 = (IsoChunkRegion)this.connectedNeighbors.get(var2);
            IsoWorldRegion var3 = var1.getIsoWorldRegion();
            if (var3 != null) {
               return var1;
            }
         }

         return null;
      }
   }
}

File: Vanilla/sources/zdoc/media/lua/shared/Library/IsoChunkRegion.lua
---@class IsoChunkRegion : zombie.iso.areas.isoregion.regions.IsoChunkRegion
---@field private manager IsoRegionManager
---@field private isInPool boolean
---@field private color Color
---@field private ID int
---@field private zLayer byte
---@field private squareSize byte
---@field private roofCnt byte
---@field private chunkBorderSquaresCnt byte
---@field private enclosed boolean[]
---@field private enclosedCache boolean
---@field private connectedNeighbors ArrayList|Unknown
---@field private allNeighbors ArrayList|Unknown
---@field private isDirtyEnclosed boolean
---@field private isoWorldRegion IsoWorldRegion
IsoChunkRegion = {}

---@public
---@param arg0 IsoChunkRegion
---@return boolean
function IsoChunkRegion:containsConnectedNeighbor(arg0) end

---@public
---@return boolean
function IsoChunkRegion:getIsEnclosed() end

---@protected
---@param arg0 IsoChunkRegion
---@return void
function IsoChunkRegion:removeConnectedNeighbor(arg0) end

---@public
---@return int
function IsoChunkRegion:getChunkBorderSquaresCnt() end

---@public
---@return void
function IsoChunkRegion:resetRoofCnt() end

---@public
---@return IsoChunkRegion
function IsoChunkRegion:getConnectedNeighborWithLargestIsoWorldRegion() end

---@public
---@return int
function IsoChunkRegion:getNeighborCount() end

---@protected
---@return void
function IsoChunkRegion:resetChunkBorderSquaresCnt() end

---@public
---@param arg0 byte
---@param arg1 boolean
---@return void
function IsoChunkRegion:setEnclosed(arg0, arg1) end

---@public
---@return int
function IsoChunkRegion:getID() end

---@protected
---@return boolean
function IsoChunkRegion:isInPool() end

---@public
---@return ArrayList|Unknown
function IsoChunkRegion:getConnectedNeighbors() end

---@protected
---@return IsoChunkRegion
function IsoChunkRegion:getFirstNeighborWithIsoWorldRegion() end

---@public
---@return Color
function IsoChunkRegion:getColor() end

---@protected
---@return void
function IsoChunkRegion:removeChunkBorderSquaresCnt() end

---@public
---@return int
function IsoChunkRegion:getRoofCnt() end

---@protected
---@return IsoChunkRegion
function IsoChunkRegion:reset() end

---@public
---@return IsoWorldRegion
function IsoChunkRegion:unlinkFromIsoWorldRegion() end

---@private
---@return void
function IsoChunkRegion:resetEnclosed() end

---@public
---@param arg0 IsoChunkRegion
---@return void
function IsoChunkRegion:addNeighbor(arg0) end

---@public
---@param arg0 IsoWorldRegion
---@return void
function IsoChunkRegion:setIsoWorldRegion(arg0) end

---@public
---@return void
function IsoChunkRegion:addSquareCount() end

---@public
---@return int
function IsoChunkRegion:getSquareSize() end

---@public
---@return ArrayList|Unknown
function IsoChunkRegion:getDebugConnectedNeighborCopy() end

---@public
---@param arg0 IsoChunkRegion
---@return void
function IsoChunkRegion:addConnectedNeighbor(arg0) end

---@protected
---@param arg0 IsoChunkRegion
---@return void
function IsoChunkRegion:removeNeighbor(arg0) end

---@protected
---@return void
function IsoChunkRegion:setDirtyEnclosed() end

---@public
---@return void
function IsoChunkRegion:addRoof() end

---@protected
---@return ArrayList|Unknown
function IsoChunkRegion:getAllNeighbors() end

---@protected
---@param arg0 int
---@param arg1 int
---@return void
function IsoChunkRegion:init(arg0, arg1) end

---@public
---@return int
function IsoChunkRegion:getzLayer() end

---@public
---@param arg0 int
---@return boolean
function IsoChunkRegion:containsConnectedNeighborID(arg0) end

---@public
---@return IsoWorldRegion
function IsoChunkRegion:getIsoWorldRegion() end

---@public
---@return void
function IsoChunkRegion:addChunkBorderSquaresCnt() end

---@protected
---@return void
function IsoChunkRegion:unlinkNeighbors() end
